Document fluent style of typed access interfaces (#626)
cap-js/docs: java/data.md

For all structured types of the CDS model, accessor interfaces can be generated using the [CDS Maven Plugin](./advanced#staticmodel). The generated accessor interfaces allow for hybrid access and easy serialization to JSON.

> By default, the accessor interfaces prodive the setter and getter methods inspired by the JavaBeans specification and the way used by the common ORM frameworks.

Reviewer: ```suggestion
By default, the accessor interfaces provide the setter and getter methods inspired by the JavaBeans specification and the way used by the common ORM frameworks.
```

Following example uses accessor interfaces that have been generated with the default (JavaBeans) style:

```java
    Authors author = Authors.create();
    author.setName("Emily Brontë");

    Books book = Books.create();
    book.setAuthor(authors);
    book.setTitle("Wuthering Heights");
```

Alternatively, you can generate accessor interfaces in _fluent style_. In this mode, the getter methods are named after the property names and the setter methods return the accessor interface itself to enable fluent chaining.

Following is the example of the fluent style:

```java
   Authors author = Authors.create().name("Emily Brontë");
   Books.create().author(author).title("Wuthering Heights");
```

The generation mode is configured by the property [`<methodStyle>`](./assets/cds-maven-plugin-site/generate-mojo.html#methodstyle) of the goal `cds:generate` of the CDS Maven Plugin and affects all entities and event contexts in your services. The default value is the `BEAN`, which represents the JavaBeans-style of interfaces.

The style of the interfaces is the matter of personal taste and you can select one that is best for you.

> The way the interfaces are generated does not affect how the entities represented by them are handled by CAP and does not change the way how the event contexts and entities delivered by the CAP looks like. Such interfaces from CAP are always modelled in the default JavaBeans style.

Reviewer: ```suggestion
The way the interfaces are generated determines only how data is accessed by custom code. It does not affect how the data is represented in memory and handled by the CAP Java runtime. 

Moreover, it does not change the way how the event contexts and entities delivered by the CAP looks like. Such interfaces from CAP are always modelled in the default JavaBeans style.
```

For all structured types of the CDS model, accessor interfaces can be generated using the [CDS Maven Plugin](./advanced#staticmodel). The generated accessor interfaces allow for hybrid access and easy serialization to JSON.

> By default, the accessor interfaces are generated so that the properties of an entity can be read and changed with the getter and the setter methods that mimic the Java Bean API or a way the properties are defined in the common ORM frameworks.

Reviewer: ```suggestion
By default, the accessor interfaces are generated so that the elements of an entity are accessed by getter and the setter methods that follow the design pattern for properties of the [JavaBeans specification](https://download.oracle.com/otndocs/jcp/7224-javabeans-1.01-fr-spec-oth-JSpec/) (8.3) as commonly used  by other ORM frameworks.
```

Author: Why mention the spec at all? We can say something like "classic" or "familiar to seasoned Java developers".

> Following is the example of the interface generated with the default style:

Reviewer: ```suggestion
Following is the example uses accessor interfaces that have generated with the default style:
```

```java

>     Authors authors = Authors.create("...");
>     authors.setName("Emily Brontë");
> 	
>     Books books = Books.create();
>     books.setAuthor(authors);
>     books.setTitle("Wuthering Heights");

Reviewer: ```suggestion
    Authors author = Authors.create();
    author.setName("Emily Brontë");
	
    Books book = Books.create();
    book.setAuthor(authors);
    book.setTitle("Wuthering Heights");
```

```

> You can also generate accessor interfaces in the style of the _Fluent Interface_ pattern. In this mode, the getter methods are named after the property names and the setter methods return the accessor interface itself to enable chained calls. 

Reviewer: ```suggestion
Alternatively, you can generate accessor interfaces in _fluent style_: In this mode, the read methods are named after the property names. The write methods return the accessor interface itself to enable chained calls. 
```

Following is the example of the fluent style:

```java

>    Authors authors = Authors.create("...").name("Emily Brontë");
>    Books.create().author(authors).title("Wuthering Heights");

Reviewer: ```suggestion
   Authors author = Authors.create().name("Emily Brontë");
   Books.create().author(author).title("Wuthering Heights");
```

```

> The generation mode is configured by the property [`<methodStyle>`](./assets/cds-maven-plugin-site/generate-mojo.html#methodstyle) of the goal `cds:generate` of the CDS Maven Plugin and affects all entities and event contexts in your services. The generation mode does not affect the way how the accessor interfaces are used by CAP in the CQN statements or in the event handlers, so you can choose the way that is suitable for you. The event contexts and accessor interfaces delivered by the CAP itself always follow the Java Bean style.

Reviewer: ```suggestion
The generation mode is configured by the property [`<methodStyle>`](./assets/cds-maven-plugin-site/generate-mojo.html#methodstyle) of the goal `cds:generate` of the CDS Maven Plugin and affects all entities and event contexts in your services. The generation mode does not affect the way how the query [model interfaces](../java/advanced#model-interfaces) of the [static model](../java/advanced#static-model-in-the-query-builder) for the query builder. The event contexts and accessor interfaces delivered by the CAP itself always follow the JavaBeans style.
```

Author: Not sure if that phrase is better. I wanted to say that the difference in style does not influence the way the framework works. That this style is the matter of taste.

Reviewer: Aha. I thought you wanted to say that the style does not affect how the "builder interfaces" (model interfaces) are generated. So we want to document:

* style of accessor interfaces is a question of personal taste
* the CAP Java runtime does not care
* model interfaces are not affected

The Maven plugin can generate the accessor interfaces in two different styles: by default the accessor interfaces are generated so that the properties of an entity can be read and changed with the getter and the setter methods that mimic the Java Bean API or a way the properties are defined in the common ORM frameworks.

Following is the example of the Java Bean style:

```java
    Authors authors = Authors.create("...");
    authors.setName("Emily Brontë");
	
    Books books = Books.create();
    books.setAuthor(authors);
    books.setTitle("Wuthering Heights");
```

You can also switch the generation mode to the _fluent_ style, which generates accessor interfaces in the style of the Fluent API pattern. In this mode, the getter methods are named after the property names and the setter methods return the accessor interface itself to enable chained calls. 

Following is the example of the fluent style:

```java
   Authors authors = Authors.create("...").name("Emily Brontë");
   Books.create().author(authors).title("Wuthering Heights");
```

> The generation mode is configured by the property `<methodStyle>` of the [CDS Maven Plugin](./advanced#staticmodel) and affects all entities and event contexts in your services. The generation mode does not affect the way how the accessor interfaces are used by CAP in the CQN statements or in the event handlers. The event contexts and accessor interfaces delivered by the CAP itself always follow the Java Bean style.

Reviewer 2: Do you want to point to the `cds-maven-plugin` documentation to the goal `generate` ? Then, I think the link is wrong.

Author: I added the direct link. I wonder how to test that it works..

Reviewer 2: Maybe locally with this capire dev. server setup.

Reviewer 2: see: https://github.com/cap-js/docs#running-locally

> The generation mode is configured by the property `<methodStyle>` of the goal [cds:generate](./assets/cds-maven-plugin-site/generate-mojo.html#methodstyle) of the CDS Maven Plugin and affects all entities and event contexts in your services. The generation mode does not affect the way how the accessor interfaces are used by CAP in the CQN statements or in the event handlers. The event contexts and accessor interfaces delivered by the CAP itself always follow the Java Bean style.

Author: They seem to be checked in the end. https://github.com/cap-js/docs/actions/runs/7530987932/job/20498548492?pr=626

With `../assets` the path looks wrong even in the browser.

> The generation mode is configured by the property `<methodStyle>` of the goal [cds:generate](./assets/cds-maven-plugin-site/generate-mojo.html#methodstyle) of the CDS Maven Plugin and affects all entities and event contexts in your services. The generation mode does not affect the way how the accessor interfaces are used by CAP in the CQN statements or in the event handlers. The event contexts and accessor interfaces delivered by the CAP itself always follow the Java Bean style.

Reviewer 2: mhh, strange, that's what I found in the existing docu.

> The generation mode is configured by the property `<methodStyle>` of the goal [cds:generate](./assets/cds-maven-plugin-site/generate-mojo.html#methodstyle) of the CDS Maven Plugin and affects all entities and event contexts in your services. The generation mode does not affect the way how the accessor interfaces are used by CAP in the CQN statements or in the event handlers. The event contexts and accessor interfaces delivered by the CAP itself always follow the Java Bean style.

Reviewer 2: can try without the anchor #methodstyle ?

::: warning

> If you use the fluent style generation and have a custom action or a function in your service, you have to explicitly set the event context as completed by explicitly calling the method `setCompleted()` in comparison to the Java Bean style where the event context is completed automatically when you call `setResult()` method. 

Author: Something like that. I stumbled upon this while playing and we do the same in our sample here https://github.com/SAP-samples/cap-sflight/blob/c868b1c1f7ebf7fbc58ceacce3335b0d4ec964cc/srv/src/main/java/com/sap/cap/sflight/processor/DeductDiscountHandler.java#L67

Thought that it is worth mentioning.

> If you use the fluent style generation and have a custom action or a function in your service, you have to explicitly set the event context as completed by explicitly calling the method `setCompleted()` in comparison to the Java Bean style where the event context is completed automatically when you call `setResult()` method. 

Reviewer 2: Ok, thanks for pointing me to the current implementation. Once we have fixed / changed the implementation we can adjust this documentation again.

> If you use the fluent style generation and have a custom action or a function in your service, you have to explicitly set the event context as completed by explicitly calling the method `setCompleted()` in comparison to the Java Bean style where the event context is completed automatically when you call `setResult()` method. 

Author: We fixed this. I could retain the warning with release-specific clause or we could merge this after 2.6.0 and omit it.

For all structured types of the CDS model, accessor interfaces can be generated using the [CDS Maven Plugin](./advanced#staticmodel). The generated accessor interfaces allow for hybrid access and easy serialization to JSON.

> By default, the accessor interfaces provide the setter and getter methods inspired by the JavaBeans specification and the way used by the common ORM frameworks.

Author: I am trying to say that by default the way objects look like is familiar to the seasoned Java developer used to JPA.

> By default, the accessor interfaces provide the setter and getter methods inspired by the JavaBeans specification and the way used by the common ORM frameworks.

Reviewer 3: Understood. So to keep it simple, we could omit this part, right? We don't really loose information that way, do we?
```suggestion
By default, the accessor interfaces provide the setter and getter methods inspired by the JavaBeans specification.
```

Following example uses accessor interfaces that have been generated with the default (JavaBeans) style:

```java
    Authors author = Authors.create();
    author.setName("Emily Brontë");

    Books book = Books.create();
    book.setAuthor(authors);
    book.setTitle("Wuthering Heights");
```

Alternatively, you can generate accessor interfaces in _fluent style_. In this mode, the getter methods are named after the property names. To enable fluent chaining, the setter methods return the accessor interface itself.

Following is an example of the fluent style:

```java
   Authors author = Authors.create().name("Emily Brontë");
   Books.create().author(author).title("Wuthering Heights");
```

The generation mode is configured by the property [`<methodStyle>`](./assets/cds-maven-plugin-site/generate-mojo.html#methodstyle) of the goal `cds:generate` provided by the CDS Maven Plugin. The selected `<methodStyle>` affects all entities and event contexts in your services. The default value is `BEAN`, which represents JavaBeans-style interfaces.

> The style of the interfaces depends on your preference. Select the one that is best for you.

Reviewer: That's not an individual preference. You would have to agree within your team which style should be used within your application.

> The style of the interfaces depends on your preference. Select the one that is best for you.

Author: And this is probably the decision you take once per project.

> The style of the interfaces depends on your preference. Select the one that is best for you.

Reviewer 3: Is this then a recommendation we want to give here? Or is this sth. a team anyways knows and agrees upon project start? If we want to give that recommendation, we can use a green box and spent one or two sentences on that to cover all aspects.

> The style of the interfaces depends on your preference. Select the one that is best for you.

Author: Both options are fine, the general recommendation could be that one should stick to the defaults. 

We just wanted to highlight that there is a choice there.

> The style of the interfaces depends on your preference. Select the one that is best for you.

Reviewer 3: ```suggestion
Once, when starting a project, select the style of the interfaces that is best for your team and project. We recommend the default JavaBeans style.
```
